packages/@react-aria/autocomplete/src/useAutocomplete.ts

```diff
@@ -0,0 +1,146 @@
+  let register = (callback) => {
+    callbackRef.current = callback;
+  };
+
+  // For textfield specific keydown operations
+  let onKeyDown = (e: BaseEvent<KeyboardEvent<any>>) => {
+    if (e.nativeEvent.isComposing) {
+      return;
+    }
+
+    // TODO: how best to trigger the focused element's action? Currently having the registered callback handle dispatching a
+    // keyboard event
+    switch (e.key) {
+      case 'Escape':
+        if (state.inputValue !== '' && !isReadOnly) {
+          state.setInputValue('');
+        } else {
+          e.continuePropagation();
+        }
+
+        break;
+      case 'Home':
+      case 'End':
+      case 'ArrowUp':
+      case 'ArrowDown':
+        // Prevent these keys from moving the text cursor in the input
+        e.preventDefault();
+        break;
+    }
+
+    if (callbackRef.current) {
+      let focusedNodeId = callbackRef.current(e);
+      state.setFocusedNodeId(focusedNodeId);
+    }
+  };
+
+  let {labelProps, inputProps, descriptionProps} = useTextField({
+    ...props as any,
+    onChange: state.setInputValue,
+    onKeyDown: chain(onKeyDown, props.onKeyDown),
+    value: state.inputValue,
+    autoComplete: 'off'
+  }, inputRef);
+
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-aria/autocomplete');
+  let menuProps = useLabels({
+    id: menuId,
+    'aria-label': stringFormatter.format('menuLabel'),
+    'aria-labelledby': props['aria-labelledby'] || labelProps.id
+  });
+
+  return {
+    labelProps,
+    inputProps: mergeProps(inputProps, {
+      'aria-haspopup': 'listbox',
+      'aria-controls': menuId,
+      // TODO: readd proper logic for completionMode = complete (aria-autocomplete: both)
+      'aria-autocomplete': 'list',
+      'aria-activedescendant': state.focusedNodeId ?? undefined,
+      // TODO: note that the searchbox role causes newly typed letters to interrupt the announcement of the number of available options in Safari.
+      // I tested on iPad/Android/etc and the combobox role doesn't seem to do that but it will announce that there is a listbox which isn't true
+      // and it will say press Control Option Space to display a list of options which is also incorrect. To be fair though, our combobox doesn't open with
+      // that combination of buttons
+      role: 'searchbox',
+      // This disable's iOS's autocorrect suggestions, since the autocomplete provides its own suggestions.
+      autoCorrect: 'off',
+      // This disable's the macOS Safari spell check auto corrections.
+      spellCheck: 'false'
+    }),
+    menuProps: mergeProps(menuProps, {
+      shouldUseVirtualFocus: true,
+      onHoverStart: (e) => {
+        // TODO: another thing to think about, what is the best way to past this to menu/wrapped collection component so that hovering on
+        // a item also updates the focusedNode. Perhaps we should just pass down setFocusedNodeId instead
+        state.setFocusedNodeId(e.target.id);
+      }
```

Comment:
Is this weird also? Focus on hover happens in the menu (and future other wrapped collections) and thus needs to communicate this change upwards. This felt better that having the menu hooks accept `setFocusedNodeId`